src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1382 @@
+
+    {
+      // Check to see if we have cached compilation data for this module.
+      auto lock = cachedData.lockShared();
+      KJ_IF_SOME(c, *lock) {
+        // We new new here because v8 will take ownership of the CachedData instance,
+        // even tho we are maintaining ownership of the underlying buffer.
+        data = new v8::ScriptCompiler::CachedData(c.begin(), c.size());
+      }
+    }
+
+    // Note that the Source takes ownership of the CachedData pointer that we pass in.
+    // Do not use data after this point.
+    v8::ScriptCompiler::Source source(externed ? js.strExtern(ptr) : js.str(ptr), origin, data);
+
+    auto maybeCached = source.GetCachedData();
+    if (maybeCached != nullptr) {
+      if (!maybeCached->rejected) {
+        options = v8::ScriptCompiler::CompileOptions::kConsumeCodeCache;
+      } else {
+        // In this case we'll just log a warning and continue on. This is potentially
+        // a signal that something with the compile cache is not working correctly but
+        // it is not a fatal error. If we spot this in the wild, it warrants some
+        // investigation.
+        LOG_WARNING_ONCE("NOSENTRY Cached data for ESM module was rejected");
+      }
+    }
+
+    v8::Local<v8::Module> module;
+
+    if (!v8::ScriptCompiler::CompileModule(js.v8Isolate, &source, options).ToLocal(&module)) {
+      return v8::MaybeLocal<v8::Module>();
+    }
+
+    // If the options are still kNoCompileOptions, then we did not have or use cached
+    // data. We should generate the cache now, if possible. We lock to ensure that
+    // we do not generate the cache multiple times needlessly. When the lock is acquired
+    // we check again to see if the cache is still empty, and skip generating if it is not.
+    if (options == v8::ScriptCompiler::CompileOptions::kNoCompileOptions) {
+      auto lock = cachedData.lockExclusive();
+      if (*lock == kj::none) {
+        auto cached = v8::ScriptCompiler::CreateCodeCache(module->GetUnboundModuleScript());
+        if (cached != nullptr) {
+          auto buf = kj::heapArray<kj::byte>(cached->length);
+          memcpy(buf.begin(), cached->data, cached->length);
+          *lock = kj::mv(buf);
+        }
+      }
+    }
+
+    return module;
+  }
+
+private:
+  v8::MaybeLocal<v8::Value> evaluate(
+      Lock& js,
+      v8::Local<v8::Module> module,
+      const CompilationObserver& observer) const override {
+    if (!ensureInstantiated(js, module, observer, this)) {
+      return v8::MaybeLocal<v8::Value>();
+    };
+    return module->Evaluate(js.v8Context());
+  }
+
+  kj::Array<const char> source;
+  kj::ArrayPtr<const char> ptr;
+  // When externed is true, the source buffer is passed into the isolate as an externalized
+  // string. This is only appropriate for built-in modules that are compiled into the binary.
+  bool externed = false;
+  kj::MutexGuarded<kj::Maybe<kj::Array<kj::byte>>> cachedData;
+};
+
+class SyntheticModule final : public Module {
```

Comment:
What is a synthetic module?

Reply:
There are generally two kinds of modules in v8: ESM and Synthetic. An ESM module is one whose module namespace is defined by the `exports` in the module code. A Synthetic module is one that is not based on ESM syntax, where the runtime essentially synthetically creates the module namespace programmatically.